exp/bufferpool/pool.go

```diff
@@ -4,49 +4,151 @@ package bufferpool
 import "sync"
 
 const (
-	nBuckets = 20
+	defaultMinSize     = 1024
+	defaultBucketCount = 10
 )
 
-// A default global pool.
-var Default Pool
+var defaultPool = &Pool{}
 
-// A pool of buffers, in variable sizes.
+// Set the global pool.  This can be used to override the default implementation
+// to optimize for different allocation profiles. For example, applications that
+// target embedded platforms may wish to lower the minimum buffer size from 1KiB
+// to 512B, or less.
+//
+// The default implementation uses BasicPool, which is suitable for a wide range
+// of applications.  Most applications will not benefit from using a custom pool,
+// so be sure to profile.
+//
+// Set panics if p == nil.
+func Set(p *Pool) {
+	if p == nil {
+		panic("Pool(nil)")
+	}
+
+	defaultPool = p
+}
+
+// Get calls .Get() on the global pool, and resizes the resulting buf
+// to size.
+func Get(size int) []byte {
+	return defaultPool.Get(size)[:size]
+}
+
+// Put calls .Put(buff) on the global pool after zeroing the supplied
+// slice. Note that only the first len(buf) bytes are zeroed, and not
+// the full backing array.
+//
+// TODO:  should we zero out the entire array?  Users may have resized
+// the slice, so it may contain sensitive data.
+func Put(buf []byte) {
+	for i := range buf {
+		buf[i] = 0
+	}
+
+	if capacity := cap(buf); capacity > 0 {
+		defaultPool.Put(buf[:capacity])
+	}
+}
+
+// Pool maintains a list of BucketCount buckets containing buffers
+// of exponentially-increasing size, starting from MinSize. Buffers
+// whose capacity exceeds MinSize << BucketCount are ignored and left
+// to the garbage-collector.
+//
+// The zero-value Pool is ready to use, defaulting to BucketCount=10
+// and MinSize=1024 (1KiB - ~1MiB buffer sizes).  Most applications
+// will not benefit from tuning these parameters.
+//
+// As a general rule, increasing MinSize reduces GC latency at the
+// expense of increased memory usage.  Increasing BucketCount can
+// reduce GC latency in applications that frequently allocate buffers
+// of size >=1MiB.
 type Pool struct {
-	buckets [nBuckets]sync.Pool
+	once                 sync.Once
+	MinSize, BucketCount int
+	buckets              []bucket
 }
 
-// Get a buffer of the given length. Its capacity may be larger than the
-// requested size.
+// Get a buffer of len >= size.  The caller will resize the
+// buffer to size, so implementations need not bother.  The
```

Comment:
The "implementations..." language should probably change now that this is not an interface.